contracts/pools/yv2/StableSwapYv2.vy

```diff
@@ -0,0 +1,996 @@
+    raise
+
+
+@view
+@internal
+def get_D_mem(rates: uint256[N_COINS], _balances: uint256[N_COINS], _amp: uint256) -> uint256:
+    xp: uint256[N_COINS] = empty(uint256[N_COINS])
+    for i in range(N_COINS):
+        xp[i] = rates[i] * _balances[i] / PRECISION
+
+    return self.get_D(xp, _amp)
+
+
+@view
+@external
+def get_virtual_price() -> uint256:
+    """
+    @notice The current virtual price of the pool LP token
+    @dev Useful for calculating profits
+    @return LP token virtual price normalized to 1e18
+    """
+    D: uint256 = self.get_D(self._xp(self._stored_rates()), self._A())
+    # D is in the units similar to DAI (e.g. converted to precision 1e18)
+    # When balanced, D = n * x_u - total virtual value of the portfolio
+    return D * PRECISION / ERC20(self.lp_token).totalSupply()
+
+
+@view
+@external
+def calc_token_amount(amounts: uint256[N_COINS], is_deposit: bool) -> uint256:
+    """
+    @notice Calculate addition or reduction in token supply from a deposit or withdrawal
+    @dev This calculation accounts for slippage, but not fees.
+         Needed to prevent front-running, not for precise calculations!
+    @param amounts Amount of each coin being deposited
+    @param is_deposit set True for deposits, False for withdrawals
+    @return Expected amount of LP tokens received
+    """
+    amp: uint256 = self._A()
+    rates: uint256[N_COINS] = self._stored_rates()
+    _balances: uint256[N_COINS] = self.balances
+    D0: uint256 = self.get_D_mem(rates, _balances, amp)
+    for i in range(N_COINS):
+        _amount: uint256 = amounts[i]
+        if is_deposit:
+            _balances[i] += _amount
+        else:
+            _balances[i] -= _amount
+    D1: uint256 = self.get_D_mem(rates, _balances, amp)
+    token_amount: uint256 = ERC20(self.lp_token).totalSupply()
+    diff: uint256 = 0
+    if is_deposit:
+        diff = D1 - D0
+    else:
+        diff = D0 - D1
+    return diff * token_amount / D0
+
+
+@external
+@nonreentrant('lock')
+def add_liquidity(
+    _amounts: uint256[N_COINS],
+    _min_mint_amount: uint256,
+    _use_underlying: bool = False
+) -> uint256:
+    """
+    @notice Deposit coins into the pool
+    @param _amounts List of amounts of coins to deposit
+    @param _min_mint_amount Minimum amount of LP tokens to mint from the deposit
+    @param _use_underlying If True, deposit underlying assets instead of cyTokens
+    @return Amount of LP tokens received by depositing
+    """
+    assert not self.is_killed
+
+    amp: uint256 = self._A()
+    rates: uint256[N_COINS] = self._current_rates()
+    _lp_token: address = self.lp_token
+    token_supply: uint256 = ERC20(_lp_token).totalSupply()
+
+    # Initial invariant
+    D0: uint256 = 0
+    old_balances: uint256[N_COINS] = self.balances
+    if token_supply != 0:
+        D0 = self.get_D_mem(rates, old_balances, amp)
+
+    # Take coins from the sender
+    coin: address = ZERO_ADDRESS
```

Comment:
moving this line to line 369 (immediately after the `if amount!= 0:`) will help with the bytecode size by freeing the memory slot when the scope closes